### DISCUSSION TITLE
with keyword & nullable [Language feature request]

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5628] [Language Ideas] [answered]
Please allow ```PotentiallyNullableVariable with {...}```
Currently we have ```with``` keyword with serveral types, in c# 10 we've introduced default nullable warning, so it makes sense to support it while variables marked with nullable is showing up a lot

**Comment by huoyaoyuan** [comment, marked as answer]
If I'm understanding correctly, do you mean:
`nullableValue with { ... } `
will become
`nullableValue is null ? null : nullableValue with { ... }`?

This looks like breaking.

**Reply by Xyncgas** [reply]
no,
Currently,
```
struct A{public int a;}
A? a;
a with {a=5}
```
isn't supported.

**Reply by Xyncgas** [reply]
it would be nice to

**Reply by HaloFour** [reply]
You might want to include those details in the original post.  Specifically that you want this to work with nullable value types and what you want it to do when that NVT happens to be `null`.

**Reply by Xyncgas** [reply]
> You might want to include those details in the original post. Specifically that you want this to work with nullable value types and what you want it to do when that NVT happens to be `null`.

I am sorry

**Reply by Xyncgas** [reply]
We all know ```T with``` returns the same type T, so it is odd that T? with doesn't return T? but just stops working

**Reply by Xyncgas** [reply]
you know what I am saying?

**Reply by HaloFour** [reply]
Unfortunately I do not, there isn't much context to that statement.

As of now `with` only works specifically with records, as those are the only types that synthesize the appropriate members.  The language team has not yet decided to open that to other constructs, such as `Nullable<T>`, which is what `T?` really is.

**Reply by 333fred** [reply]
> As of now with only works specifically with records

`with` works with all struct types. However, `Nullable<T>` does not have any writable properties, so while you _can_ with it (https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD6wQfgAQR4C8eMqANuQNwCwAUAPrEF4DuUCAFngN54C+VIA===), you can't actually do anything useful with that `with`. You could potentially raise an API request with the BCL to add an `init` to `Nullable<T>.Value`, which would allow you do `a with { Value = ... }`, but honestly I don't think that makes much sense as an API pattern. You really want to do something like `a = a is null ? null : a.Value with { ... }`.

**Reply by jnm2** [reply]
Maybe we can have a `with?` keyword, because the conditional syntax is going to hurt. Especially when it breaks a chain like this:

```cs
expr with { A.NullableValue.C = 42 }
```

Ooh:
```cs
expr with { A.NullableValue?.C = 42 }
```

**Reply by 333fred** [reply]
I would only want to see something like that if https://github.com/dotnet/csharplang/issues/2883 were implemented.

**Reply by jnm2** [reply]
I would greatly love to see (or help) #2883 be implemented. That would have 2-3 orders of magnitude more benefit for me personally compared to this. And both would be interesting.
